clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -295,24 +297,27 @@ private long sendProducerData(long now) {
     }
 
     private boolean maybeSendTransactionalRequest(long now) {
-        TransactionManager.TxnRequestHandler nextRequestHandler = transactionManager.nextRequestHandler();
-        if (nextRequestHandler == null) {
-            log.trace("TransactionalId: {} -- There are no pending transactional requests to send",
-                    transactionManager.transactionalId());
-            return false;
-        }
+        String transactionalId = transactionManager.transactionalId();
+        if (transactionManager.isCompletingTransaction() && !transactionManager.hasPartitionsToAdd() && accumulator.hasUnflushedBatches()) {
+            if (transactionManager.isAborting())
+                accumulator.abortUnsentBatches(new KafkaException("Failing batch since transaction was aborted"));
 
-        if (nextRequestHandler.isEndTxn() && transactionManager.isCompletingTransaction() && accumulator.hasUnflushedBatches()) {
             if (!accumulator.flushInProgress())
```

Comment:
A comment to the effect of 'There may be batches in the queue which were previously sent and are in retry. We have to flush those to keep the sequence numbers on the client and servers in sync'.